contracts/DKVDaggerHashimoto.sol

```diff
@@ -212,7 +216,66 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
             uint256 parent = uint256(hash0) % rows;
             uint256 kvIdx = parent + (startShardId << shardEntryBits);
             bytes memory data = maskedData[i];
-            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data), "invalid access proof");
+            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data),
+                    "invalid access proof");
+
+            assembly {
+                mstore(data, hash0)
+                hash0 := keccak256(data, add(maxKvSize, 0x20))
+                mstore(data, maxKvSize)
+            }
+        }
+        return hash0;
+    }
+
+    /* Helper function to avoid stack to deep error */
+    function _generateKVIdx(
+        bytes32 hash0,
+        uint256 startShardId,
+        uint256 shardLenBits
+    ) internal view returns (uint256) {
+        uint256 rows = 1 << (shardEntryBits + shardLenBits);
+        uint256 parent = uint256(hash0) % rows;
+        uint256 kvIdx = parent + (startShardId << shardEntryBits);
+        return kvIdx;
+    }
+
+    function _generateChunkIdx(
+        bytes32 hash0
+    ) internal view returns (uint256) {
+        require((CHUNK_SIZE != 0) && ((CHUNK_SIZE & (CHUNK_SIZE - 1)) == 0), "CHUNK_SIZE invalid");
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+        uint256 chunksPerKVBlock = maxKvSize / CHUNK_SIZE;
+        /* If maxKvSize is smaller than CHUNK_SIZE, early exit */
+        if (chunksPerKVBlock == 0) return 0;
+        uint256 chunkIdx = uint256(hash0) % chunksPerKVBlock;
+        return chunkIdx;
+    }
+
+    /*
+     * Run a modified hashimoto hash,
+     * with Merkle inclusion proofs
+     */
+    function _hashimotoMerkleProof(
+        uint256 startShardId,
+        uint256 shardLenBits,
+        bytes32 hash0,
+        bytes32[][] memory proofsDim2,
+        bytes[] memory maskedData
+    ) internal view returns (bytes32) {
+        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
+        require(proofsDim2.length == randomChecks, "proofs vs checks: length mismatch");
+        require(maskedData.length <= CHUNK_SIZE, "calldata too large");
+        uint256 maxKvSize = 1 << maxKvSizeBits;
+
+        for (uint256 i = 0; i < randomChecks; i++) {
+            require(maskedData[i].length == maxKvSize, "invalid proof size");
+            uint256 kvIdx = _generateKVIdx(hash0, startShardId, shardLenBits);
+            uint256 chunkIdx = _generateChunkIdx(hash0);
```

Comment:
I guess this may be wrong - a kvIdx = 12 will always access chunk 12 (suppose max value size is 128KB), where we want the access to be uniform.

Instead, we should have

`
uint256 rows = 1 << (shardEntryBits + shardLenBits + chunkLenBits);
uint256 parent = uint256(hash0) % rows;
uint256 chunkIdx = parent + (startShardId << (shardEntryBits + chunkLenBits));
uint256 kvIdx = chunkIdx >> chunkLenBits;

Reply:
We haven't changed too much. The caculation method of kvIdx persists:
function _generateKVIdx(
        bytes32 hash0,
        uint256 startShardId,
        uint256 shardLenBits
    ) internal view returns (uint256) {
        uint256 rows = 1 << (shardEntryBits + shardLenBits);
        uint256 parent = uint256(hash0) % rows;
        uint256 kvIdx = parent + (startShardId << shardEntryBits);
        return kvIdx;
    }
It is uniformally distributed as before. 
The tricky part is chunkIdx. Here is story of what I understand:
Now we have hierarchy : shard - KV block - chunks
A shard has multiple KV blocks (say 128KB per block), a block has 32 chunks (say 4K per chunk). 
If we count everything in chunk, yes, term `rows` represents how many chunks we focus and use hash0 to get an offset (parent), then a base position added --- we are done. 

However if we examine the API of MerkleLib.getProof, bu design, it accepts a chunkIdx **within** the a KV Block, where in our example above, it should be range from 0 to 31. 

So if we caculate:
uint256 chunkIdx = parent + (startShardId << (shardEntryBits + chunkLenBits));
it still needs:
chunkIdx = chunkIdx % (1 <<chunkLenBits ) 
               = (parent + (startShardId << (shardEntryBits + chunkLenBits)) ) % ( 1 <<chunkLenBits )
               = parent % (1 << chunkLenBits)
               = uint256(hash0) % (1 << (shardEntryBits + shardLenBits + chunkLenBits)) % (1 << chunkLenBits)
               = uint256(hash0) % (1 << chunkLenBits)
since (1 << chunkLenBits) is a factor of (1 << (shardEntryBits + shardLenBits + chunkLenBits));
That's my function _generateChunkIdx.

uint256 kvIdx = chunkIdx >> chunkLenBits
                      =  ( parent + (startShardId << (shardEntryBits + chunkLenBits)) ) >> chunkLenBits
                      =  ( uint256(hash0) % (1 << (shardEntryBits + shardLenBits + chunkLenBits)) >> chunkLenBits ) +      
                          (startShardId << shardEntryBits)
 where uint256(hash0) % (1 << (shardEntryBits + shardLenBits)) =/=  ( uint256(hash0) % (1 << (shardEntryBits + shardLenBits + chunkLenBits)) >> chunkLenBits, can't quite tell which one is more evenly distributed. 

My naive thinking is their Lebesgue measure should be same.............

Reply:
At least can we make sure chunkIdx calculation is correct..........

Reply:
What I mean here is not `P(chunkIdx)`, where chunkIdx is uniform-distributed in [0,31], I mean `P(chunkIdx | kvIdx)` is not uniform-distributed, e.g., if kvIdx = 13, then `P(chunkIdx = 13 | kvIdx = 13) = 1`, while `P(chunkIdx \neq 13 | kvIdx = 13) = 0`.

The key here is the lowest bits are reused in both chunkIdx and kvIdx, which does not provide sufficient entry as chunkIdx and kvIdx correlated.

Reply:
That makes sense!
I didn't consider conditional possibility. You are right.